### ISSUE TITLE
externalize model configurations

### ISSUE THREAD

**Issue by behrica** [zmedelis/bosquet#14]
While I was looking into fixes for #8 #9 
and getting OpenAI and Azure OpenAI working 
and assuming that prompt templates should be shareable across users

I think that an different, more flexible, approach is needed to separate prompts and the model(s) setup (which has indeed secrete keys which should not go into code)
I  want to be able to have a **fixed** template such as:
```
{:text-analyzer/summarize-to-sentence
 "Provide a summary of the following {{text-type|default:text}} in a single sentence:

Text:
{{text}}

Summary:
{% llm-generate max-tokens=60 var-name=completion %}"
 }
```

and  a user of this template can "configure" which model (and which model setup including keys)
are used without changing the template as such.

This can IMHO not work via using environement variables. For different models with different setups this is impossible.

**Comment by behrica** [comment]
Maybe a flexible way to do this, is to assume that parameters to `llm-generate` can be as well pass when calling `gen/complete` , for example:

```clojure
(gen/complete-template 
{:text "my text to summarize"}
{:llm-generate {:impl :azure
                        :api-key (my-read-key-function)}})
```

**Comment by behrica** [comment]
For a template like this;§

```clojure
{:role            "As a brilliant {{you-are}} answer the following question."
    :question        "What is the distance between Io and Europa?"
    :question-answer "Question: {{question}}  Answer: {% llm-generate var-name=answer %}"
    :self-eval       "{{answer}} Is this a correct answer? {% llm-generate var-name=test model=text-curie-001 %}"}
    ```
 Maybe the above need to be in this form:
 ```
 (gen/complete-template 
{:text "my text to summarize"}
{[ :question-answer :llm-generate] 
   {:impl :azure :api-key (my-read-key-function)}
 [ :self-eval :llm-generate] 
   {:impl :azure :api-key (my-read-key-function)}})
 ```   
 This would allow as well to have different modles for each prompt, and support future tags, appart from `llm-generate`

**Comment by behrica** [comment]
I did an implementation which would be use dlike this:
```clojure
(def synopsis
  (gen/complete-template
   synopsis-template
   {:title "Mr. X"
    :genre "crime"}
   {:api-key "sk-xxxxxxxx"}))
```   
for OpenAI, and

````clojure
(def sentiments (gen/complete-template sentimental
                                       {:text-type "tweets" :tweets tweets}
                                       {:model "testtextdavanci003"
                                        :api-key "xxxxx"
                                        :api-endpoint "https://zzzzzz.openai.azure.com"
                                        :impl :azure}))
```` 

for Azure OpenAI .

I removed as well teh check for kez in environemnt or a file.
With above code , te user can implement whatever he likes how to get the keys .

This fixes #9 

I can do  a PR, if you atr interested

**Comment by behrica** [comment]
This code still makes one assumption, namely that even in a complex template all calls to `llm-generate`  use the same mode setup.
And it assumes that we only have a single 'tag', which need to be configured.

I started to experiment how to get a new tag function , which uses an embeddings model to find "similar" text + a vector database.
In this case there would be an "other model to setup". So such template would use 2 different models, which needed to be configured in some form seperately.
Or an other complex template might want to use different completion models at different places.

**Comment by zmedelis** [comment]
> I started to experiment how to get a new tag function

More tag functions will definitely be needed.

Please do a draft PR and let's work on this together to get to a final implementation.

**Comment by behrica** [comment]
#15 created.
I added as well quite some discussion in it, as I was not sure what flexibility we need to "configure" the calls to `llm-generate`.
I think we need to have a way that different calls to `llm-generate` in different places of a prompt template can be configured externally and different (outside the prompt template) likely via merging of internal and external options maps. 

This would support to have different models used in different positions of a prompt template. (and this need to support as well composition of prompt templates)

I am not sure about the best way to do this and if selmer provides eventually some mechanisms to do this.

**Comment by behrica** [comment]
solved by #15
